### ISSUE TITLE
Test becomes flaky by upgrading pytest from 7.4.4 to 8.1.0 -> fails normally but is successfull when run in verbose more

### ISSUE THREAD

**Issue by mkashwin** [pytest-dev/pytest#12071]
- [x] a detailed description of the bug or problem you are having
  In my github actions, I am seeing a strange behaviour of one of my unit tests failing when executed normally but when executed in debug mode or with -v ( verbose mode) the test passes. In the normal mode, it appears that my application cache / a object level dict attribute is being cleared, causing Key error later on.  Details of the test case and the logs from the test runs on github runner are provided below. This was not the behaviour with pytest 7.4.4
 

- [x] output of `pip list` from the virtual environment you are using
   The poetry lock file with all dependencies is [here](https://github.com/mkashwin/unifiednamespace/blob/main/05_sparkplugb/poetry.lock)

- [x] pytest and operating system versions
   Ubuntu 22.04.4 LTS ( as provided in the github runner) 
   pytest version 8.1.0

- [x] minimal example if possible
  The link to the unit test case which is failing is 
  [test_spb2unspublisher](https://github.com/mkashwin/unifiednamespace/blob/4d616a857ed0fe225a88812594fbcb58ea572877/05_sparkplugb/test/test_spb2unspublisher.py#L168)

  [Run1 : done without debug mode and fails](https://github.com/mkashwin/unifiednamespace/actions/runs/8136886960/job/22234054817)
   - The error here  ` KeyError: 'group1_/edge_node_1/MyDevice'`  makes no sense as the key was explicitly set in the actual code code
      [Application Code](https://github.com/mkashwin/unifiednamespace/blob/4d616a857ed0fe225a88812594fbcb58ea572877/05_sparkplugb/src/uns_spb_mapper/spb2unspublisher.py#L203) with the logic to set the cache key 

  [Run3 : Done with debug mode and no further changes and passes](https://github.com/mkashwin/unifiednamespace/actions/runs/8136886960/job/22244913210)

**Comment by The-Compiler** [comment]
Can you try downgrading to 7.4.4 again and double-check it indeed passes there? This looks entirely unrelated to pytest...

**Comment by mkashwin** [comment]
This is working fine on 7.4.4.

**Comment by RonnyPfannschmidt** [comment]
nm, i missunderstood the nesting setup

**Comment by mkashwin** [comment]
Issue also present in pytest 8.1.1 which was released last night

**Comment by mkashwin** [comment]
@Zac-HD Apologies for not giving that earlier, I though giving the link to the github project with direct reference to test case and code being tested would have made life easier. 

Looks like 8.1.0 got yanked and now I am unable to reproduce this with 8.1.1.
I tried to create a minimal reproducible example but am unable to reproduce this with 8.1.1
Even when I try downgrading to 8.1.0 it is not reproducing.
Closing the issue for now, will reopen if this surfaces again along with the minimal reproducible example

Thanks for your efforts
